### PR TITLE
[InstCombine] Update the `select` operand when the `cond` is `trunc` and has the `nuw` or `nsw` property. 

### LINKED ISSUE #96765: trunc nuw x to i1 as branch/select cond implies x
https://alive2.llvm.org/ce/z/grTNkc
```
----------------------------------------
define i32 @src(i32 %x, i32 %y) {
#0:
  %trunc = trunc nuw i32 %x to i1
  %ret = select i1 %trunc, i32 %x, i32 %y
  ret i32 %ret
}
=>
define i32 @tgt(i32 %x, i32 %y) {
#0:
  %trunc = trunc nuw i32 %x to i1
  %ret = select i1 %trunc, i32 1, i32 %y
  ret i32 %ret
}
Transformation seems to be correct!

```
```

----------------------------------------
declare void @f()

define i32 @src2(i32 %x, i32 %y) {
bb1:
  %cond = trunc nuw i32 %x to i1
  br i1 %cond, label %bb2, label %bb3

bb2:
  call void @f()
  br label %bb3

bb3:
  %ret = phi i32 [ %y, %bb2 ], [ %x, %bb1 ]
  ret i32 %ret
}
=>
declare void @f()

define i32 @tgt2(i32 %x, i32 %y) {
bb1:
  %cond = trunc nuw i32 %x to i1
  br i1 %cond, label %bb2, label %bb3

bb2:
  call void @f()
  br label %bb3

bb3:
  %ret = phi i32 [ %y, %bb2 ], [ 0, %bb1 ]
  ret i32 %ret
}
Transformation seems to be correct!

Summary:
  2 correct transformations
  0 incorrect transformations
  0 failed-to-prove transformations
  0 Alive2 errors
```

### PR DESCRIPTION
**Pull request by c8ef**
This patch updates the select operand when the cond has the nuw or nsw property. Considering the semantics of the nuw and nsw flag, if there is no poison value in this expression, this code assumes that X can only be 0, 1 or -1.

close: #96765
alive2: https://alive2.llvm.org/ce/z/3n3n2Q

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 1]
```diff
@@ -4201,5 +4201,15 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
     }
   }
 
+  // select (trunc nuw X to i1), X, Y --> select (trunc nuw X to i1), 1, Y
+  if (auto *TI = dyn_cast<TruncInst>(CondVal)) {
+    Value *Trunc;
+    if (TI->hasNoUnsignedWrap() && match(CondVal, m_Trunc(m_Value(Trunc))) &&
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:4207]:
```suggestion
    if (match(CondVal, m_NUWTrunc(m_Value(Trunc))) &&
```

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 2]
```diff
@@ -4201,5 +4201,15 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
...
+        match(TrueVal, m_Specific(Trunc))) {
+      return SelectInst::Create(
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:4209]: Use replaceOperand() instead, to preserve select metadata.

**Review by nikic — COMMENTED** [review]
Please add an alive2 proof to the patch description.

Note that the same transform also works for trunc nsw and -1, though that's probably less practically relevant.

**Comment by nikic** [comment]
> Note that the same transform also works for trunc nsw and -1, though that's probably less practically relevant.

There's also the variant where the trunc operand is used in FalseVal rather than TrueVal. In that case we can replace with zero.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 3]
```diff
@@ -4201,5 +4201,23 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
     }
   }
 
+  // select (trunc nuw X to i1), X, Y --> select (trunc nuw X to i1), 1, Y
+  // select (trunc nuw X to i1), Y, X --> select (trunc nuw X to i1), Y, 0
+  // select (trunc nsw X to i1), X, Y --> select (trunc nsw X to i1), -1, Y
+  // select (trunc nsw X to i1), Y, X --> select (trunc nsw X to i1), Y, 0
+  Value *Trunc;
+  if (match(CondVal, m_NUWTrunc(m_Value(Trunc)))) {
+    if (match(TrueVal, m_Specific(Trunc)))
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:4210]:
```suggestion
    if (TrueVal == Trunc)
```
etc.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 4]
```diff
@@ -4201,5 +4201,23 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
...
+      return replaceOperand(SI, 1, ConstantInt::get(TrueVal->getType(), 1));
+    if (match(FalseVal, m_Specific(Trunc)))
+      return replaceOperand(SI, 2, ConstantInt::get(TrueVal->getType(), 0));
+  }
+  if (match(CondVal, m_NSWTrunc(m_Value(Trunc)))) {
+    if (match(TrueVal, m_Specific(Trunc)))
+      return replaceOperand(SI, 1, ConstantInt::get(TrueVal->getType(), -1));
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:4217]:
```suggestion
      return replaceOperand(SI, 1, Constant::getAllOnesValue(TrueVal->getType()));
```
Please add a test using i128 for this, which your current implementation miscompiles due to incorrect extension.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 1 above)
Comment (c8ef) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 2 above)
Comment (c8ef) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Done.

**Comment by c8ef** [comment]
> Please add an alive2 proof to the patch description.
> 
> Note that the same transform also works for trunc nsw and -1, though that's probably less practically relevant.

Added all four transformations and relevant alive2 proof.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 3 above)
Comment (c8ef) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Fixed.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 4 above)
Comment (c8ef) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: I didn't even notice that, thanks for guiding me!

**Review by nikic — APPROVED** [review]
LGTM

This could be generalized using simplifyWithOpReplaced, but probably not worthwhile. The non-trunc variant of that is: https://github.com/llvm/llvm-project/blob/99b85cae628c1cc5641944290712cd84ccf1f6c8/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp#L3691-L3702

**Comment by c8ef** [comment]
Would it be better to place this transformation at https://github.com/llvm/llvm-project/blob/99b85cae628c1cc5641944290712cd84ccf1f6c8/llvm/lib/Analysis/InstructionSimplify.cpp#L4812? If yes, I'd like to relocate the implementation there.

**Comment by nikic** [comment]
> Would it be better to place this transformation at
> 
> https://github.com/llvm/llvm-project/blob/99b85cae628c1cc5641944290712cd84ccf1f6c8/llvm/lib/Analysis/InstructionSimplify.cpp#L4812
> ? If yes, I'd like to relocate the implementation there.

The transform can't go into InstSimplify because it modifies/creates an instruction.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request merged**